include/circt/Dialect/FIRRTL/FIRRTLStructure.td

```diff
@@ -144,6 +144,57 @@ def FExtModuleOp : FIRRTLOp<"extmodule",
   let hasVerifier = 1;
 }
 
+def FMemModuleOp : FIRRTLOp<"memmodule",
+      [IsolatedFromAbove, Symbol, HasParent<"CircuitOp">, OpAsmOpInterface,
+       DeclareOpInterfaceMethods<FModuleLike>,
+       DeclareOpInterfaceMethods<HWModuleLike>]> {
+  let summary = "FIRRTL Generated Module";
+  let description = [{
+    The "firrtl.memmodule" operation represents an external reference to a
+    memory module. See the "firrtl.mem" op for a deeper explantation of the
+    parameters.
+    
+    A "firrtl.mem" operation is typically lowered to this operation when they
+    are not directly lowered to an implementation by the compiler.
+  }];
+  let arguments =
+    (ins UI32Attr:$numReadPorts, UI32Attr:$numWritePorts,
+         UI32Attr:$numReadWritePorts, TypeAttr:$dataType, UI32Attr:$maskBits,
+         UI32Attr:$readLatency, UI32Attr:$writeLatency, UI64Attr:$depth,
+         ArrayAttr:$extraPorts, AnnotationArrayAttr:$annotations);
+  let results = (outs);
+  let regions = (region AnyRegion:$body);
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    OpBuilder<(ins "StringAttr":$name, "ArrayRef<PortInfo>":$ports,
+                   "uint32_t":$numReadPorts,  "uint32_t":$numWritePorts,
+                   "uint32_t":$numReadWritePorts, 
+                   "FIRRTLType":$dataType, "uint32_t":$maskBits,
+                   "uint32_t":$readLatency, "uint32_t":$writeLatency,
+                   "uint64_t":$depth,
+                   CArg<"ArrayAttr", "ArrayAttr()">:$annotations)>
+  ];
+
+  let extraClassDeclaration = [{
+    /// Get the width of this memory's data type.
+    size_t dataWidth() {
+      return dataType().cast<FIRRTLType>().getBitWidthOrSentinel();
```

Comment:
Are we expecting no aggregate data type here, we can use this to get the width of aggregate types also, 
 
 ```suggestion
       return getBitWidth(dataType().cast<FIRRTLType>()).getValue();
```
But I guess, the data type must be flattened when we lower. So we could add a verifier for the dataype or constraint it to `UInt` !